http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/http2/Http2ToHttpInboundAdapter.java

```diff
@@ -56,6 +63,30 @@ protected void channelRead0(ChannelHandlerContext ctx, Http2Frame frame) throws
         }
     }
 
+    @Override
+    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+        if ((cause instanceof ReadTimeoutException || cause instanceof WriteTimeoutException) && ctx.channel().parent() != null) {
+            healthCheck(ctx);
+        }
+
+        ctx.fireExceptionCaught(cause);
+    }
+
+    private void healthCheck(ChannelHandlerContext ctx) {
+        Channel parent = ctx.channel().parent();
+        log.debug(() -> "ReadTimeoutException occurred, sending PING frame to " + parent);
+        parent.attr(KEEP_ALIVE).set(false);
```

Comment:
Is there a race between this and getting a stream from the `MultiplexChannelRecord#acquire`? For example, is it possible for `acquire()` to think the channel is still live when it might not be?

Reply:
Yes, it's possible another request comes in right before `KEEP_ALIVE` is set to false and picks up this channel, but since we are shutting down the child channel very soon, that request should fail fast.

Reply:
Wouldn't it be possible that it can take up to the read timeout in some cases?

Reply:
Oh actually yes, I was only thinking for the case where it would fail to write the PING frame and then would get shut down. I'll think it a bit more to how to better address it.

(I forgot to mention that we will reduce the read timeout for certain service clients. I'll make that change.)

Reply:
We can add another keep alive flag check for parent channel in `HealthCheckedChannelPool` and that should reduce the chances. 

`channel.parent() != null && !channel.parent().attr(KEEP_ALIVE).get()`. 

What do you think?